File: 209_MinimumSizeSubarraySum.py
#49.86%
class Solution(object):
    def minSubArrayLen(self, s, nums):
        """
        :type s: int
        :type nums: List[int]
        :rtype: int
        """
        res, past, start = len(nums) + 1, 0, 0
        for i in xrange(len(nums)):
            past += nums[i]
            if past >= s:
                while past - nums[start] >= s:
                    past -= nums[start]
                    start += 1
                res = min(res, i - start + 1)
                past -= nums[start]
                start += 1
        if res == len(nums) + 1:
            return 0
        return res

File: 209_MinimumSizeSubarraySum.cpp
//50.66%
//A DP problems
class Solution {
public:
    int minSubArrayLen(int s, vector<int>& nums) {
        if(nums.empty())
			return 0;
		int res(nums.size() + 1), start(0), past(0);
		for(int i = 0; i < nums.size(); ++i){
			past += nums[i];
			if(past >= s){
				while(past - nums[start] >= s)
					past -= nums[start++];
				res = min(res, i - start + 1);
				past -= nums[start++];
			}
		}
		if(res == nums.size() + 1)
			return 0;
		return res;
    }
};